Appendix B - `operationId` and `description` guidelines for MCP and AI Agent readiness (#664)
camaraproject/Commonalities: documentation/CAMARA-API-Design-Guide.md

- If the subject can be identified from the access token and the optional [`device` object | `phoneNumber` field](*) is also included in the request, then the server will return an error with the `422 UNNECESSARY_IDENTIFIER` error code. This will be the case even if the same [ device | phone number ](*) is identified by these two methods, as the server is unable to make this comparison.
```

> ## Appendix B (Informative): `operationId` and `description` Guidelines for MCP and AI Agent Readiness

Reviewer: Because this appendix targets MCP/AI-agent readiness, it's worth remembering that descriptions contribute significantly to the token length of tool definitions, and every tool definition competes for the model's context window. Long, templated descriptions across dozens of operations add up quickly and can make it hard to justify including them at all.

Should try to avoid restating in the description what the model can already derive from the tool's schemas. Of the five template fields, two fall into this category:

- `Requires` largely duplicates the tool's `inputSchema`, which is always present and describes the required inputs.
- `Returns` overlaps with the response schema / `outputSchema` if there is one.

I'd suggest the description templates lead with the parts that aren't recoverable from the schema and treat `Requires`/`Returns` as optional, to be used only when they add meaning beyond the schema.

- `operationId` MUST be present on every operation.
- `operationId` MUST be in lowerCamelCase format and MUST NOT exceed 64 characters in length. This ensures compatibility with most code generators and vendor tooling limits. (These requirements are formally expressed by the regular expression: `^[a-z][a-zA-Z0-9]{0,63}$`.)
- `operationId` MUST follow the shape `<verb><Noun>[<Qualifier>]`, for example:
  - `createSession` (verb + noun)
  - `getDeviceLocation` (verb + compound noun)
  - `deleteSessionById` (verb + noun + qualifier)

  The `<Noun>` represents the core resource, entity, or concept being acted upon — and MAY be a compound noun (e.g., `DeviceLocation`, `UserProfile`) formed from multiple domain terms.

  The `<Qualifier>` is optional and, when used, MUST serve only to disambiguate two operations that would otherwise share the same `<verb><Noun>`. When the qualifier disambiguates on an identifying parameter, it SHOULD take the form `By<Param>` (e.g., `deleteSessionById`). Otherwise, it SHOULD be a single PascalCase noun or short noun phrase appended directly to `<Noun>`.

> - `operationId` MUST NOT be prefixed with the HTTP method name, with the exception of `get` when used as an ordinary English verb.

Reviewer 2: `delete` is also an HTTP method name, is on the approved CRUD verb list, and appears in this appendix's own example (`deleteSessionById`) — as written, the rule forbids its own examples. The actual method-leakage problem is `post`/`put`/`patch`: across the published camaraproject repos only ~5 path operations are affected (the `postTrafficInfluence`/`patchTrafficInfluence` family and `postServiceCapability`), while `get` (57 operations) and `delete` (34) are the two most common verbs in the catalog.

```suggestion
- `operationId` MUST NOT be prefixed with `post`, `put`, or `patch`. (`get` and `delete` are ordinary English verbs describing the operation's action and are not considered HTTP-method leakage.)
```

Note: a further ~25 `post*Notification` operationIds exist across the catalog, but all are `callbacks` definitions following the CAMARA notification template — implemented by the API consumer and never exposed as tools. Suggest stating in the B.1 intro that these rules apply to path operations only, with `callbacks` operationIds out of scope.

  This avoids duplicating information already present in the OpenAPI `path`/`method` pair and adds no disambiguation value for automated selection.

> - `operationId` MUST NOT include a version marker** such as `V1`, `_v1`, `v2`, or similar. Versioning belongs in the API path (e.g., `/v1/sessions`) or headers, not in the operation identifier.

Reviewer 2: ```suggestion
- `operationId` MUST NOT include a version marker such as `V1`, `_v1`, `v2`, or similar. Versioning belongs in the API path (e.g., `/v1/sessions`) or headers, not in the operation identifier.
```

> - Within a single API document, designers SHOULD use one consistent read verb (e.g., always `get` rather than alternating between `get`, `retrieve`, and `fetch`) to promote clarity and reduce cognitive load for consumers.

Reviewer 2: Agree with the intent, but the example labels the Design Guide's own pattern as the anti-pattern. Scanning the published camaraproject repos: `fetch` is used by no API at all, and of the 8 API documents that mix read verbs, 6 do so systematically following [6.5](#65-post-or-get-for-transferring-sensitive-or-complex-data) — `get*` for `GET` resource reads and `retrieve*` for `POST /retrieve-*` queries, with the operationId mirroring the mandated verb path (e.g. `getSession` + `retrieveSessionsByDevice` in QualityOnDemand). In CAMARA, `retrieve` is not a synonym of `get`; it marks the POST-based query style. The genuine inconsistencies are of a different kind: one API inverts the pairing (`retrieveOperation` on a `GET`, `getEsimProfileStatus` on a `POST /retrieve-status`), and a few APIs use `read` where the catalog norm is `get`. The most extensive case is DedicatedNetworks, which uses `read*` consistently across all four of its documents (`readNetwork`, `readAccess`, `readNetworkServiceArea`, `readNetworkProfile`/`readNetworkProfiles`) while correctly using `retrieve*` for its POST-based query — aligning it to `get` would mean renaming five operationIds, which is breaking for generated clients and underlines that this appendix needs an applicability statement (new APIs / new major versions only).

```suggestion
- Within a single API document, designers SHOULD use one consistent read verb per retrieval style: `get` for reading resources via `GET`, and `retrieve` for POST-based queries as defined in [6.5. POST or GET for Transferring Sensitive or Complex Data](#65-post-or-get-for-transferring-sensitive-or-complex-data), mirroring the `retrieve-*` verb path. Synonyms such as `read` or `fetch` SHOULD NOT be used in their place.
```

> - `operationId` SHOULD be unique within the document to prevent collisions in code generation, tool registries, and AI agent tool-calling systems.

Reviewer 2: SHOULD is weaker than what already applies today: the [OpenAPI 3.0 spec](https://spec.openapis.org/oas/v3.0.3#operation-object) requires that `operationId` is unique within a document (MUST), and the CAMARA linting rule `operation-operationId-unique` already reports a duplicate as an error ([Linting-rules.md](https://github.com/camaraproject/Commonalities/blob/main/documentation/Linting-rules.md)). The published repos contain exactly one duplicate today: WebRTC uses `createNotificationChannelSubscription` twice, once for the path operation and once for its notification callback. The linter does not report it because it only checks path operations. If the rules in this appendix apply only to path operations (see comment on the HTTP-method rule above), this case is no longer relevant.

The collision described in #623 ("duplicates cause tool collision") is a different one: the same `operationId` used in different APIs. This already happens today. `createSession` is used by QualityOnDemand, SessionInsights and WebRTC with three different meanings; DedicatedNetworks and MultiPointVPN both use `createNetwork` and `deleteNetwork`; DedicatedNetworks and InHomeDeviceManagement both use `listDevices`; and four APIs kept the placeholder names from the subscription template (`retrieveSubscription`, `retrieveSubscriptionList`, `deleteSubscription`) instead of adding an API-specific noun. Uniqueness within one document does not prevent these collisions — the second bullet below addresses them.

```suggestion
- `operationId` MUST be unique within the API document, as required by the OpenAPI specification and enforced by the CAMARA linting rule `operation-operationId-unique`.
- The `<Noun>` SHOULD be unambiguous across CAMARA APIs (e.g. `retrieveSimSwapSubscription` rather than a generic `retrieveSubscription`): when several APIs are exposed through one MCP server or tool registry, generic names collide. Full cross-API uniqueness (e.g. via namespace prefixes) is a tool-definition concern and out of scope for this document.
```

> #### Proposal: Approved Verb List

Reviewer 2: Three suggestions for this section, based on a scan of all path operations in the published camaraproject repos:

**1. Base the CRUD core list on actual CAMARA usage, not on generic naming.**
The verbs CAMARA APIs use today are: `get` (62 operations), `retrieve` (54), `create` (42), `delete` (41), `update` (16), `list` (10). This is already a clean CRUD core, so the list should be: `create`, `get`, `retrieve`, `list`, `update`, `delete`. Two changes compared to the proposal:
- `retrieve` belongs in the core list with its own meaning (POST-based queries per section 6.5, see the comment on the read-verb rule above) — not under "`get` — Replaces: `retrieve`".
- `replace` is used by no CAMARA API today. It can be added later when a real need appears.

**2. Action verbs: add `check` and `verify`; start with a shorter list.**
`check` (7 APIs, e.g. `checkSimSwap`) and `verify` (4 APIs, e.g. `verifyLocation`) are the central verbs of CAMARA's verification and fraud-prevention APIs — they must be on the list. In the other direction, 13 of the 21 proposed action verbs (`approve`, `publish`, `archive`, `restore`, `deactivate`, `import`, `export`, `search`, `preview`, `transfer`, `unassign`, `tag`, `untag`) are used by no CAMARA API. Suggest starting with the verbs that have real usage and extending the list over time through the review process — a shorter list is easier to agree on and to maintain.

**3. "Forbidden … never permitted" is stronger than the rule it belongs to.**
The section intro says verbs "SHOULD be restricted", so "never permitted" does not fit. Suggest renaming to "**Discouraged verbs** (SHOULD NOT be used)". The content itself looks reasonable — for the record, only `manage` (1 operation) and `perform` (1 operation) appear in the catalog today, and no `get*Data` name exists.

One more small point: "MUST undergo a formal review process" — please name where this happens (for example: a Commonalities issue), otherwise API designers cannot follow the rule.

Reviewer 3: Let me add related comments:

1.- In the CRUD Core Verb List, `retrieve` is listed as a synonym to be replaced by get:

`get` — Fetch a single resource by identity. Replaces: fetch, retrieve, read, find.

This directly contradicts section 6.5 of the same Design Guide, which mandates `retrieve` as the verb for POST-based queries (POST /retrieve-*). In CAMARA, `retrieve` is not a synonym of `get`; it marks a distinct retrieval style. There are currently 54 operations across the catalog using retrieve with this specific meaning.
Suggested correction: move `retrieve` to the CRUD Core list as a standalone entry:

`retrieve` — POST-based query, as defined in §6.5. Mirrors the retrieve-* verb path.

And remove it from the "Replaces" list of get.

2.- two verbs central to CAMARA's verification and fraud-prevention APIs are missing from the list:

`check` — used by 7 APIs (e.g., checkSimSwap)
`verify` — used by 4 APIs (e.g., verifyLocation)

Suggestion: add `check` and `verify` to the Action Verbs list.

Reviewer 2: @rartych I unresolved/reopended this discussion as the comments are only partially applied, especially are `check` and `verify` still not part of the Action Verb list. Is there a reason for that?

- `replace` — Full replacement of a resource (PUT semantics). Replaces: `put`, `overwrite`.
- `delete` — Remove a resource permanently. Replaces: `remove`, `destroy`, `drop`.

**Action Verbs** (for non-CRUD operations):
- `send` — Dispatch a message, notification, or payload.
- `submit` — Hand off for processing (e.g., forms, orders, applications).
- `approve` — Grant explicit authorization.
- `cancel` — Abort an in-progress or future operation.
- `publish` — Make a resource publicly available.
- `archive` — Move to long-term / read-only storage.
- `restore` — Reverse an archive or soft delete.
- `enable` / `disable` — Toggle active state.
- `activate` / `deactivate` — Lifecycle state transitions.
- `import` / `export` — Bulk data movement across system boundaries.
- `search` — Intentional query with user-supplied terms (distinct from `list`, which implies browsing).
- `validate` — Check correctness without side effects.
- `preview` — Dry-run or read-only rendering of an action’s outcome.
- `transfer` — Move ownership or location between entities.
- `assign` / `unassign` — Attach or detach a resource to an owner.
- `tag` / `untag` — Apply or remove a label.

> **Forbidden Verbs** (never permitted):

Reviewer 3: Suggestion for Line 1639: **Discouraged Verbs (SHOULD NOT be used)** 
(Up to Community consideration) 

To be aligned with the wording used in Line 1611

Verbs used in `operationId` SHOULD be restricted to the list below. Verbs not on this list MUST undergo a formal review process before use.

- `process` — too vague; does not convey what the operation does.
- `handle` — internal/implementation language, not semantic.
- `manage` — a catch-all with no meaning to an agent.
- `do`, `run`, `execute`, `perform`, `trigger` — no accompanying noun can make these specific enough.
- `get*Data` patterns (e.g., `getInvoiceData`) — the `Data` suffix adds noise; use `getInvoice` instead.

### B.2. Operation `description` Completeness Rules

Automated consumers (e.g., AI agents, MCP clients) typically evaluate operations using only the operation object — its `description`, `parameters`, `requestBody`, and `responses` — not the global `info.description`. To ensure they can make correct choices, each operation must be self-contained for the purpose of selection and invocation.

Therefore:

> - `operation description` MUST be present and non-empty.

Reviewer 3: ```suggestion
- operation `description` MUST be present and non-empty.
```

Reviewer 3: suggestion to remark the concept according to yaml definition.
Other approach could be `operation.description`

> - `operation description` MUST NOT be verbatim identical to `summary`.

Reviewer 3: ```suggestion
- operation `description` MUST NOT be verbatim identical to `summary`.
```
